hail/python/hail/linalg/blockmatrix.py

```diff
@@ -2169,14 +2043,75 @@ def _svd_gramian(self, compute_uv):
 block_matrix_type.set(BlockMatrix)
 
 
-def _shape(b):
-    if isinstance(b, int) or isinstance(b, float):
-        return 1, 1
-    if isinstance(b, np.ndarray):
-        b = _ndarray_as_2d(b)
+def _is_scalar(x):
+    return isinstance(x, float) or isinstance(x, int)
+
+
+def _shape_after_broadcast(left, right):
+    """
+    Follows numpy's strategy of broadcasting through right-align shapes and
+    compare corresponding dimensions. See:
+    https://docs.scipy.org/doc/numpy-1.15.0/user/basics.broadcasting.html#general-broadcasting-rules
+    """
+    def join_dim(l_size, r_size):
+        if not (l_size == r_size or l_size == 1 or r_size == 1):
+            raise ValueError(f'Incompatible shapes for broadcasting: {left}, {right}')
+
+        return max(l_size, r_size)
+
+    def pad(arr, n):
+        return [1 for _ in range(n)] + arr
+
+    diff_len = len(left) - len(right)
+    if diff_len < 0:
+        left = pad(left, -diff_len)
+    elif diff_len > 0:
+        right = pad(right, diff_len)
+
+    return [join_dim(l, r) for l, r in zip(left, right)]
+
+
+# Can easily be transformed to hl.tensor()
```

Comment:
I'd prefer to not have this comment because I don't think it conveys anything that isn't obvious from the function name `_to_bmir`

Reply:
Fair enough, would have been better as note in the PR rather than code